src/J2N/IO/CharReadOnlyMemoryAdapter.cs

```diff
@@ -0,0 +1,193 @@
+        }
+
+        public override CharBuffer Duplicate() => Copy(this);
+
+        public override char Get()
+        {
+            if (position == limit)
+            {
+                throw new BufferUnderflowException();
+            }
+            return sequence.Span[position++];
+        }
+
+        public override char Get(int index)
+        {
+            if ((uint)index >= (uint)limit)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index));
+            }
+            return sequence.Span[index];
+        }
+
+        public override sealed CharBuffer Get(char[] destination, int offset, int length) // J2N TODO: API - rename startIndex instead of offset
+        {
+            if (destination is null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.destination);
+            if (offset < 0)
+                ThrowHelper.ThrowArgumentOutOfRange_MustBeNonNegative(offset, ExceptionArgument.offset);
+            if (length < 0)
+                ThrowHelper.ThrowArgumentOutOfRange_MustBeNonNegative(length, ExceptionArgument.length);
+            if (offset > destination.Length - length) // Checks for int overflow
+                ThrowHelper.ThrowArgumentOutOfRange_IndexLengthString(offset, length);
+            if (length > Remaining)
+                throw new BufferUnderflowException();
+
+            int newPosition = position + length;
+            sequence.Slice(position, length).CopyTo(destination.AsMemory(offset, length));
+            position = newPosition;
+            return this;
+        }
+
+        public override CharBuffer Get(Span<char> destination) // J2N specific
+        {
+            int length = destination.Length;
+            if (length > Remaining)
+                throw new BufferUnderflowException();
+
+            int newPosition = position + length;
+            sequence.Span.Slice(position, length).CopyTo(destination);
+            position = newPosition;
+            return this;
+        }
+
+        //public override bool IsDirect => false;
+
+        public override bool IsReadOnly => true;
+
+        public override ByteOrder Order => ByteOrder.NativeOrder;
+
+        protected override char[] ProtectedArray
+        {
+            get { throw new NotSupportedException(); }
+        }
+
+        protected override int ProtectedArrayOffset
+        {
+            get { throw new NotSupportedException(); }
+        }
+
+        protected override bool ProtectedHasArray => false;
+
+        public override CharBuffer Put(char value)
+        {
+            throw new ReadOnlyBufferException();
+        }
+
+        public override CharBuffer Put(int index, char value)
+        {
+            throw new ReadOnlyBufferException();
+        }
+
+        public override sealed CharBuffer Put(char[] source, int offset, int length) // J2N TODO: API - rename startIndex instead of offset
+        {
+            if (source is null)
```

Comment:
I'm curious what the point is of having these guard clauses if it's just going to throw later if it gets past them. Should these methods not just only `throw new ReadOnlyBufferException();`?

Reply:
That is the way it is done upstream: https://github.com/apache/harmony/blob/trunk/classlib/modules/nio/src/main/java/common/java/nio/CharSequenceAdapter.java#L133-L153.

Although, I commented out the extra exceptions and there are no tests verifying they are thrown. So, I am not sure why they exist.

The JDK source code is generated, so it can't be analyzed directly. I wasn't able to work out from the templates where these adapters are generated from (or whether they even exist).

https://github.com/openjdk/jdk8/tree/master/jdk/src/share/classes/java/nio

But since this is a minor thing, I will merge now and revisit later.